// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option go_package = "github.com/pingcap/tidb/planner/core";
option java_multiple_files = true;
option java_package = "io.grpc.examples.helloworld";
option java_outer_classname = "HelloWorldProto";

package join_order;

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}
  rpc SayHelloAgain(HelloRequest) returns (HelloReply){}
}

service JoinOrder{
  rpc TestJoinNode(LogicalJoinNode) returns(HelloReply){}
  rpc GetAction(State) returns(HelloReply){}
  rpc IsDQN(Empty) returns (BoolValue){}
}
// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}

message Condition{
    string funcname = 1;
    repeated string args = 2;
}

message LogicalNode{
  oneof node{
    LogicalJoinNode join_node = 1;
    LogicalTableNode table_node = 2;
  }
  repeated LogicalNode childrens = 3;
}
message LogicalJoinNode{
  string tp = 1;
  string join_type = 2;
  repeated LogicalJoinNode join_childrens = 3;
  repeated LogicalTableNode table_childrens = 4;
  repeated Condition conditions = 5;
}

message LogicalTableNode{
  string tp = 1;
  string table_name = 2;
  repeated Condition conditions = 3;
}

message State{
  LogicalNode current_join_tree = 1;
  repeated LogicalNode actions = 2;
  string original_sql = 3;
}

message Empty{ 
}

message BoolValue{
  bool value = 1;
}